### PR TITLE
[c10][rocm] fix __assert_fail() declaration mismatch error

### PR DESCRIPTION
**Pull request by shintaro-iwasaki**
Summary:
This patch fixes a compilation error in PyTorch with ROCm when `NDEBUG` is passed.

## Problem

Forward declaration of `__host__ __device__ __assert_fail()` is used in `c10/macros/Macros.h` for HIP compilation when `NDEBUG` is set  However, HIP has  `__device__ __assert_fail()` in `hip/amd_detail/amd_device_functions.h`, causing a function type error.

This issue does not appear in ROCm CI tests since it happens only when `NDEBUG` is passed.

## Solution

To solve this compilation error, this patch disables `CUDA_KERNEL_ASSERT()`, which uses `__assert_fail()`

when [EDIT] ROCm/HIP is used.
1. ~`c10/macros/Macros.h` is included for `*.hip` (precisely speaking, `__HIP__` or `__HIP_ARCH__` is defined), and~
2. ~`NDEBUG` is passed.~

~Note that there's no impact on default compilation because, without a special compilation flag, those HIP files are compiled without `-NDEBUG`. And that's why this issue has not been found.~

### Justification
[1] We cannot declare one host-and-device function for two separate host and device functions.
```c
__device__ int func() {return 0};
__host__ int func() {return 0};
// Compile error (hipcc)
// __device__ __host__ int func();
```
[2] Forward declaration of a correct `__device__` only `__assert_fail()` for `__HIP__` causes the following error:
```
pytorch/c10/util/TypeCast.h: error: reference to __device__ function '__assert_fail' in __host__ __device__ function
      ERROR_UNSUPPORTED_CAST
      ^
pytorch/c10/util/TypeCast.h: note: expanded from macro 'ERROR_UNSUPPORTED_CAST'
#define ERROR_UNSUPPORTED_CAST CUDA_KERNEL_ASSERT(false);
                               ^
pytorch/c10/macros/Macros.h: note: expanded from macro 'CUDA_KERNEL_ASSERT'
    __assert_fail(
```

[3] Maybe there's a way to properly define `__assert_fail()` for HIP + NDEBUG, but this might be too much. Please let me just disable it.

### Technical details

Error
```
pytorch/c10/macros/Macros.h: error: __host__ __device__ function '__assert_fail' cannot overload __device__ function '__assert_fail'
    __assert_fail(
    ^
/opt/rocm/hip/include/hip/amd_detail/amd_device_functions.h: note: previous declaration is here
void __assert_fail(const char *assertion,
```

CUDA definition (9.x) of `__assert_fail()`
```c
#elif defined(__GNUC__)
extern __host__ __device__ __cudart_builtin__ void __assert_fail(
  const char *, const char *, unsigned int, const char *)
  __THROW;
```

ROCm definition (the latest version)
```c
// https://github.com/ROCm-Developer-Tools/hipamd/blob/2b59661f3e51060338e0cd1e488a03b1aefdeb0c/include/hip/amd_detail/amd_device_functions.h#L1172-L1177
extern "C" __device__ __attribute__((noinline)) __attribute__((weak))
void __assert_fail(const char *assertion,
                   const char *file,
                   unsigned int line,
                   const char *function);
```

Test Plan:
CI + reproducer
```sh
python3 tools/amd_build/build_amd.py
python3 setup.py develop --cmake-only
cmake -DHIP_HIPCC_FLAGS_RELEASE="-DNDEBUG" build
cmake --build build
```

Differential Revision: D34310555

### CODE REVIEW

**Comment by xw285cornell** [comment]
cc. @jeffdaily

**Comment by shintaro-iwasaki** [comment]
@ithunnair-amd Thanks for looking at this issue. I used ROCm 4.5.0 and 4.5.2.

**Comment by mruberry** [comment]
This seems fine but I suggest adding a comment explaining the ROCm logic

File: c10/macros/Macros.h [hunk 1]
```diff
@@ -331,8 +331,12 @@ constexpr uint32_t CUDA_THREADS_PER_BLOCK_FALLBACK = 256;
 // CUDA_KERNEL_ASSERT checks the assertion
 // even when NDEBUG is defined. This is useful for important assertions in CUDA
 // code that would otherwise be suppressed when building Release.
+//
+// Declaration of __assert_fail() does not match that of ROCm (even >= 4.1),
+// so CUDA_KERNEL_ASSERT() is disabled for HIP code when NDEBUG is set.
 #if defined(__ANDROID__) || defined(__APPLE__) || defined(__XROS__) || \
-    (defined(USE_ROCM) && ROCM_VERSION < 40100)
+    (defined(USE_ROCM) && ROCM_VERSION < 40100) ||                     \
+    (defined(NDEBUG) && (defined(__HIP_ARCH__) || defined(__HIP__)))
```
Comment (xw285cornell) [inline c10/macros/Macros.h:339]: Honestly this feels like HIP is just not ready for the NDEBUG flag. How about just change it to USE_ROCM, and when it's ready we'll turn it back on? What do you think @jeffdaily @jithunnair-amd

Comment (jeffdaily) [reply c10/macros/Macros.h:339]: The PR that enabled kernel asserts for ROCm (#49624) should be partially reverted, yes.  Partial, meaning update these lines here to disable kernel asserts for ROCm, without version condition.  But the original PR made improvements to the unit tests; the tests should be skipped now but keep the other improvements.  Unfortunately, the unit tests that would exercise the kernel asserts only execute for the "slow" category of tests, and ROCm does not have a GitHub Action for "slow" at this time.  Further, perhaps related to this current PR and the incorrect function declaration, we're seeing other issues with kernel asserts in pytorch starting in ROCm 4.5.2 and through 5.0.  The feature should be disabled until we can fully revisit.

Note that we used to set NDEBUG flag for all ROCm builds until that PR.

Comment (shintaro-iwasaki) [reply c10/macros/Macros.h:339]: Okay, then I will disable `CUDA_KERNEL_ASSERT()` for ROCm. Please let me update the PR.

Comment (shintaro-iwasaki) [reply c10/macros/Macros.h]: Thanks for your comments. I updated the commit. Now it passes all tests including the ROCm ones (https://github.com/pytorch/pytorch/runs/5363652718?check_suite_focus=true).

(I'd appreciate any comments/reviews from you: @jeffdaily, @jithunnair-amd)

**Review by jeffdaily — APPROVED** [review]
LGTM

**Pull request closed without merging**